### PR TITLE
add gears.table.cycle_value

### LINKED ISSUE #2869: Missing gears.table.iterate_value in awful.widget.layoutlist example
Output of `awesome --version`:

awesome v4.3 (Too long)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

**How to reproduce the issue:**

I pasted the center popup example from https://awesomewm.org/doc/api/classes/awful.widget.layoutlist.html into a file and `require`d it in my config. When i press mod4+space awesome throws an error.

**Actual result:**

    ~/.config/awesome/layoutlist.lua: 54: attempt to call a nil value (field 'iterate_value')

**Expected result:**

The popups appears, no errors are thrown.

**Additional notes:**

I found a [post](https://www.reddit.com/r/unixporn/comments/akbu6k/awesomewm_wait_why_did_i_do_this_again/) by @Elv13 in /r/unixporn which includes a [config](https://gist.github.com/Elv13/b679fe3cb100cd63a7082cc808d316b8) with a definition of `gears.table.iterate_value`. If this function is not supposed to be part of the awesome library, maybe the code should be included in the example?

### PR DESCRIPTION
**Pull request by SethBarberee**
Fixes #2869 

* Copied over the definition from [here](https://gist.github.com/Elv13/b679fe3cb100cd63a7082cc808d316b8). 
* Added tests to go along with it

### CODE REVIEW

File: spec/gears/table_spec.lua [hunk 1]
```diff
@@ -42,6 +42,20 @@ describe("gears.table", function()
...
+            local f = gtable.iterate_value(t, "a")
+            assert.is.same(f, "b")
+        end)
+        it("with step size", function()
+            local t = { "a", "b", "c", "d" }
+            local f = gtable.iterate_value(t, "a", 2)
+            assert.is.same(f, "c")
+        end)
+    end)
+
```
Comment (actionless) [inline spec/gears/table_spec.lua]: i think it would be useful to add one more testcase using a `filter` argument, to make the coverage bot happy

Comment (SethBarberee) [reply spec/gears/table_spec.lua]: Yeah, I can add that

**Comment by Elv13** [comment]
Hi, thanks for this. 2 "minor" things:

* Can you think of a name that doesn't contain "iterate" (cycle?) ? It might confuse some people who expect this to be a loop iterator.
* Can you add an example using the `text` example template (easy) or an extended sequence template (probably overkill and hard)? The current `gears.table` is missing a lot of examples. This is mostly because it predates the example framework. Given that you are familiar with that code, I think it is acceptable to ask for this. Otherwise I will have to do it at some point, but I think I can spend my doc time on more lacking components.

File: lib/gears/table.lua [hunk 2]
```diff
@@ -245,6 +245,42 @@ function gtable.iterate(t, filter, start)
...
+-- @staticfct gears.table.iterate_value
+function gtable.iterate_value(t, value, step_size, filter, start_at)
+    local k = gtable.hasitem(t, value, true, start_at)
+    if not k then return end
+
+    step_size = step_size or 1
+    local new_key = gmath.cycle(#t, k + step_size)
+
+    if filter and not filter(t[new_key]) then
+        for i=1, #t do
```
[the remark this answers is not in the document]
Comment (SethBarberee) [reply lib/gears/table.lua]: Agreed on both

**Comment by SethBarberee** [comment]
> Hi, thanks for this. 2 "minor" things:
> 
>     * Can you think of a name that doesn't contain "iterate" (cycle?) ? It might confuse some people who expect this to be a loop iterator.

`cycle_value`?

**Comment by Elv13** [comment]
> cycle_value?

sounds good. And thanks for all that work https://github.com/awesomeWM/awesome/pull/2514 was really going nowhere. It has been on my git stash for like 5 years...

**Comment by SethBarberee** [comment]
> Very nice addition!
> 
> Thanks for your work on `gears.table`, The `gears` module is, IMO, the worst regarding documentation. I always go to the code to understand what function do what instead of reading the doc...
> 
> As @Elv13 suggested, I would also change the name too. I'm waiting an iterator function to give me [the iterator](https://en.wikipedia.org/wiki/Iterator) not to iterate by itself and give the result.

Agreed, I probably will go through add some more examples/documentation. Sounds like the best use that I can contribute.

**Comment by Aire-One** [comment]
Looks good to me! :)

Just a quick note: I would have moved the example to the function description instead of the global module description.

Anyway, thanks for your work on this one  :+1:

File: lib/gears/table.lua [hunk 3]
```diff
@@ -245,6 +245,42 @@ function gtable.iterate(t, filter, start)
...
+-- `first_index` has to be specified.
+--
+-- @tparam table t The input table.
+-- @param value A value from the table.
+-- @tparam[opt=1] number step_size How many element forward (or backward) to pick.
+-- @tparam[opt=nil] function filter An optional function. When it returns
+--  `false`, the element are skipped until a match if found. It takes the value
+--  as its sole parameter.
+-- @tparam[opt=1] number start_at Where to start the lookup from.
+-- @return The value. If there is a filter function and no element match,
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/gears/table.lua]: done

**Pull request merged**